internal/allocator/allocator.go

```diff
@@ -269,31 +272,84 @@ func (a *Allocator) AllocateFromPool(svc string, serviceIPFamily ipfamily.Family
 }
 
 // Allocate assigns any available and assignable IP to service.
-func (a *Allocator) Allocate(svc string, serviceIPFamily ipfamily.Family, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
-	if alloc := a.allocated[svc]; alloc != nil {
-		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
+func (a *Allocator) Allocate(key string, svc *v1.Service, serviceIPFamily ipfamily.Family, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
+	pinnedPriorityPools, pinnedNoPriorityPools := a.getPinnedPools(svc)
+	// If pinned pools matches on service namespace and/or its labels, then try to allocate ip from a pool based on its
+	// priority. When pools with priority is no more available, then use pool with no priority set.
+	sort.Slice(pinnedPriorityPools, func(i, j int) bool {
+		return pinnedPriorityPools[i].ServiceAllocations.Priority <
+			pinnedPriorityPools[j].ServiceAllocations.Priority
+	})
+	for _, pool := range pinnedPriorityPools {
+		if ips, err := a.AllocateFromPool(key, serviceIPFamily, pool.Name, ports, sharingKey, backendKey); err == nil {
+			return ips, nil
+		}
+	}
+	for _, pool := range pinnedNoPriorityPools {
+		if ips, err := a.AllocateFromPool(key, serviceIPFamily, pool.Name, ports, sharingKey, backendKey); err == nil {
+			return ips, nil
+		}
+	}
+	if alloc := a.allocated[key]; alloc != nil {
+		if err := a.Assign(key, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
 		return alloc.ips, nil
 	}
 
-	for poolName := range a.pools {
-		if !a.pools[poolName].AutoAssign {
+	for poolName := range a.pools.PoolsByName {
+		if !a.pools.PoolsByName[poolName].AutoAssign || a.pools.PoolsByName[poolName].ServiceAllocations != nil {
 			continue
 		}
-		if ips, err := a.AllocateFromPool(svc, serviceIPFamily, poolName, ports, sharingKey, backendKey); err == nil {
+		if ips, err := a.AllocateFromPool(key, serviceIPFamily, poolName, ports, sharingKey, backendKey); err == nil {
 			return ips, nil
 		}
 	}
 
 	return nil, errors.New("no available IPs")
 }
 
+func (a *Allocator) getPinnedPools(svc *v1.Service) ([]*config.Pool, []*config.Pool) {
```

Comment:
Can we have this return a single sorted list with the ones with priority on top and the ones without priority after? That will make the caller easier I think